wallet-new/docs/updates.md

```diff
@@ -0,0 +1,563 @@
+                     -- ^ The update has been fully applied.
+                     | Downloaded !Sha256Checksum !FilePath
+                     -- ^ The update has been downloaded but not applied yet. A Checksum
+                     -- can be stored to verify the integrity of the download.
+                     | Skipped
+                     -- ^ The update has been skipped by the user.
+                     | Proposed
+                     -- ^ The update has been just proposed and is waiting for users' feedback.
+
+## Determining the severity<a id="sec-5-2" name="sec-5-2"></a>
+
+For the proposed schema to succeed, we would need a function like the following:
+
+    updateSeverity :: UpdateProposal -> UpdateSeverity
+
+Such function would gather evidence from an `UpdateProposal` that this is a non-negotiable update.
+
+## User stories<a id="sec-5-3" name="sec-5-3"></a>
+
+In order to check the validity of the proposed plan, let's try to run trough some of the common
+user stories for the update system.
+
+-   (1) As a Daedalus user, I want the wallet to prompt me if there is a new update available.
+
+At startup Daedalus would make a call to `GET /api/v1/update` to fetch the last available update. Note
+how this endpoint would **always** return a `WalletSoftwareUpdate`.
+In case of a `WalletSoftwareUpdate` which has the `wsuUpdateState` field set to `Applied`, Daedalus would
+consider itself updated. In case the fetched `WalletSoftwareUpdate` had `Mandatory` severity (but is not
+yet applied), Daedalus would freeze the UI and prevent the user from doing anything but update.
+At this stage, closing Daedalus won't have any effect, as upon the next startup the RESTful call would be
+performed again.
+
+**N.B** A very fair point could be made that we are giving clients too many responsibility and we are offloading
+too much logic to them. If that's the case, we can easily create a new endpoint such as `GET /api/v1/update/severity`
+which would reply with a JSON `{ "updateSeverity": "mandatory" }` in case an update is available, is not fully
+applied and is marked as `Mandatory` (names and payloads might vary to make them more fitting).
+
+-   (1.1) As a Daedalus user, I want to be able to download an update.
+
+After having grabbed a `WalletSoftwareUpdate` via `GET /api/v1/update` Daedalus is now in the position to
+react to user's input. In this story, the user would choose to continue with the update process. To do so, the update
+would need to be downloaded. More specifically, the user would call `POST /api/v1/update/download` which
+would begin the download process. No state change would be necessary in the backend yet. Story `1.2.` assumes
+that this endpoint has been called and it details its behaviour.
+
+-   (1.2) As a Daedalus user, I want to be able to see the download progress for an update.
+
+To accomplish this, one possible idea would be to use [chunked transfer encoding](https://en.wikipedia.org/wiki/Chunked_transfer_encoding)
```

Comment:
I'm not sure that this is straightforward to implement with Servant. I believe that it's _possible_, but I'm afraid it might require some deep digging into darkest corners of `Servant.Server.Internal` and lots of suffering. I'm happy to be proven wrong, though.

If I'm right, we might do this in dead simple way of providing a separate endpoint for checking download progress. We also might think about something like websocket notifications, but I'm not sure if frontend guys would be happy with such complications.

Reply:
Returning chunked http responses it's a pretty standard practice for a web server, but I appreciate might not fall into Servant's standard use case. I'm happy to investigate how painful it is, and I think that if we could use a chunked transfer encoding (unless @domenkozar thinks it's a terrible idea) that would be a win, because we wouldn't need to keep any piece of separate state to check the progress of a download (and thus no separate endpoint) and we could easily treat the download operation as atomic: it's either completed in a single HTTP call, or is not. Which makes keeping the internal state consistent very easy.

Reply:
I'm aware of benefits provided by chunked encoding, I just believe that we shouldn't finalize the update api spec until we are sure that we won't have to go to hell and back to implement it.

This won't stop this particular PR to be merged, though. I think that the right way of resolving this is to create separate issue for investigating how chunked encoding can be implemented with Servant, and changing the spec with another PR if it turns out to be really hard.